winml/lib/Api/LearningModelSession.cpp

```diff
@@ -28,119 +28,114 @@ static const GUID WINML_PIX_EVAL_CAPTURABLE_WORK_GUID = __uuidof(guid_details::W
+    WINML_THROW_IF_FAILED(model->ModelEnsureNoFloat16());
+  }
+  return model.detach();
 }
 
 void LearningModelSession::Initialize() {
   // Begin recording session creation telemetry
   _winmlt::TelemetryEvent session_creation_event(
-    _winmlt::EventCategory::kSessionCreation);
+      _winmlt::EventCategory::kSessionCreation);
   // Get the optimized model proto from the learning model
-  com_ptr<winmla::IModelProto> model_proto; 
-  model_proto.attach(GetOptimizedModel());
+  com_ptr<WinML::IModel> model;
+  model.attach(GetOptimizedModel());
 
   // Create the session builder
   auto device_impl = device_.as<winmlp::LearningModelDevice>();
+  auto model_impl = model_.as<winmlp::LearningModel>();
 
-  com_ptr<winmla::IWinMLAdapter> adapter;
-  WINML_THROW_IF_FAILED(OrtGetWinMLAdapter(adapter.put()));
+  engine_factory_.copy_from(model_impl->GetEngineFactory());
 
-  com_ptr<winmla::IOrtSessionBuilder> session_builder;
-  WINML_THROW_IF_FAILED(adapter->CreateOrtSessionBuilder(
-    device_impl->GetD3DDevice(), 
-    device_impl->GetDeviceQueue(),
-    session_builder.put()));
+  com_ptr<WinML::IEngineBuilder> engine_builder;
+  engine_factory_->CreateEngineBuilder(engine_builder.put());
 
-  Ort::SessionOptions options(nullptr);
-  WINML_THROW_IF_FAILED(session_builder->CreateSessionOptions(options.put()));
+  if (device_impl->IsCpuDevice() == false) {
+    engine_builder->SetD3D12Device(device_impl->GetD3DDevice());
+    engine_builder->SetID3D12CommandQueue(device_impl->GetDeviceQueue());
+  }
 
   // Make onnxruntime apply the batch size override, if any
-  if (session_options_ && session_options_.BatchSizeOverride() != 0)
-  {
-    Ort::ThrowOnError(Ort::GetApi().AddFreeDimensionOverride(
-      options, 
-      onnx::DATA_BATCH, 
-      session_options_.BatchSizeOverride()));
+  if (session_options_ && session_options_.BatchSizeOverride() != 0) {
+    engine_builder->SetBatchSizeOverride(session_options_.BatchSizeOverride());
   }
 
-  com_ptr<winmla::IInferenceSession> session;
-  WINML_THROW_IF_FAILED(session_builder->CreateSession(
-      options, session.put(), &cached_execution_provider_));
+  com_ptr<WinML::IEngine> engine;
+  WINML_THROW_IF_FAILED(engine_builder->CreateEngine(engine.put()));
 
   // Register the custom operator registry
-  auto model = model_.as<winmlp::LearningModel>();
-  operatorRegistry_.reset(model->GetOperatorRegistry());
-  WINML_THROW_IF_FAILED(session->RegisterCustomRegistry(operatorRegistry_.get()));
+  operator_registry_ = MLOperatorRegistry(model_impl->GetOperatorRegistry(), [](auto registry) { registry->Release(); });
+  WINML_THROW_IF_FAILED(engine->RegisterCustomRegistry(operator_registry_.get()));
 
-  // Register only the transformers not already in ORT
-  session->RegisterGraphTransformers();
+  // Register transformers - this should probably not be exposed on IEngine, but an internal call as this configuration step is ort specific.
+  engine->RegisterGraphTransformers();
```

Comment:
move into onnxruntimeengine implementation?

Reply:
future refactor...

---
In reply to: [370318155](https://github.com/microsoft/onnxruntime/pull/2891#discussion_r370318155) [](ancestors = 370318155)